### ISSUE TITLE
Randomness of GATConv after initialization.

### ISSUE THREAD

**Issue by gao462** [pyg-team/pytorch_geometric#1894]
## ❓ Questions & Help

I am trying to reproduce GATConv for some other usage, but I found there are some extra randomness on att_l and att_r (for 1.6.3, if it is 1.6.1, even lin_l and lin_r have unknown randomness) after initialization even if I directly copy the GATConv code from PyTorch Geometric repo.

Below is my code, I work on CPU, use constant weight parameters, use PyTorch 1.7 set_deterministic to control scatter_add (in softmax part), but it still has a slight difference on gradients of att_l and att_r.

P.S. Also, before update to latest 1.6.3, att_r will not get any gradient although it is invovled in forward computation, but after update to 1.6.3, it will receive gradient.

```python
from typing import Union, Tuple, Optional
from torch_geometric.typing import (OptPairTensor, Adj, Size, NoneType,
                                    OptTensor)

import torch
import torch_geometric
from torch import Tensor
import torch.nn.functional as F
from torch.nn import Parameter, Linear
from torch_sparse import SparseTensor, set_diag
from torch_geometric.nn.conv import MessagePassing
from torch_geometric.utils import remove_self_loops, add_self_loops, softmax

from torch_geometric.nn.inits import glorot, zeros


class GATConv(MessagePassing):
    r"""The graph attentional operator from the `"Graph Attention Networks"
    <https://arxiv.org/abs/1710.10903>`_ paper
    .. math::
        \mathbf{x}^{\prime}_i = \alpha_{i,i}\mathbf{\Theta}\mathbf{x}_{i} +
        \sum_{j \in \mathcal{N}(i)} \alpha_{i,j}\mathbf{\Theta}\mathbf{x}_{j},
    where the attention coefficients :math:`\alpha_{i,j}` are computed as
    .. math::
        \alpha_{i,j} =
        \frac{
        \exp\left(\mathrm{LeakyReLU}\left(\mathbf{a}^{\top}
        [\mathbf{\Theta}\mathbf{x}_i \, \Vert \, \mathbf{\Theta}\mathbf{x}_j]
        \right)\right)}
        {\sum_{k \in \mathcal{N}(i) \cup \{ i \}}
        \exp\left(\mathrm{LeakyReLU}\left(\mathbf{a}^{\top}
        [\mathbf{\Theta}\mathbf{x}_i \, \Vert \, \mathbf{\Theta}\mathbf{x}_k]
        \right)\right)}.
    Args:
        in_channels (int or tuple): Size of each input sample. A tuple
            corresponds to the sizes of source and target dimensionalities.
        out_channels (int): Size of each output sample.
        heads (int, optional): Number of multi-head-attentions.
            (default: :obj:`1`)
        concat (bool, optional): If set to :obj:`False`, the multi-head
            attentions are averaged instead of concatenated.
            (default: :obj:`True`)
        negative_slope (float, optional): LeakyReLU angle of the negative
            slope. (default: :obj:`0.2`)
        dropout (float, optional): Dropout probability of the normalized
            attention coefficients which exposes each node to a stochastically
            sampled neighborhood during training. (default: :obj:`0`)
        add_self_loops (bool, optional): If set to :obj:`False`, will not add
            self-loops to the input graph. (default: :obj:`True`)
        bias (bool, optional): If set to :obj:`False`, the layer will not learn
            an additive bias. (default: :obj:`True`)
        **kwargs (optional): Additional arguments of
            :class:`torch_geometric.nn.conv.MessagePassing`.
    """
    _alpha: OptTensor

    def __init__(self, in_channels: Union[int, Tuple[int, int]],
                 out_channels: int, heads: int = 1, concat: bool = True,
                 negative_slope: float = 0.2, dropout: float = 0.,
                 add_self_loops: bool = True, bias: bool = True, **kwargs):
        kwargs.setdefault('aggr', 'add')
        super(GATConv, self).__init__(node_dim=0, **kwargs)

        self.in_channels = in_channels
        self.out_channels = out_channels
        self.heads = heads
        self.concat = concat
        self.negative_slope = negative_slope
        self.dropout = dropout
        self.add_self_loops = add_self_loops

        if isinstance(in_channels, int):
            self.lin_l = Linear(in_channels, heads * out_channels, bias=False)
            self.lin_r = self.lin_l
        else:
            self.lin_l = Linear(in_channels[0], heads * out_channels, False)
            self.lin_r = Linear(in_channels[1], heads * out_channels, False)

        self.att_l = Parameter(torch.Tensor(1, heads, out_channels))
        self.att_r = Parameter(torch.Tensor(1, heads, out_channels))

        if bias and concat:
            self.bias = Parameter(torch.Tensor(heads * out_channels))
        elif bias and not concat:
            self.bias = Parameter(torch.Tensor(out_channels))
        else:
            self.register_parameter('bias', None)

        self._alpha = None

        self.reset_parameters()

    def reset_parameters(self):
        glorot(self.lin_l.weight)
        glorot(self.lin_r.weight)
        glorot(self.att_l)
        glorot(self.att_r)
        zeros(self.bias)

    def forward(self, x: Union[Tensor, OptPairTensor], edge_index: Adj,
                size: Size = None, return_attention_weights=None):
        # type: (Union[Tensor, OptPairTensor], Tensor, Size, NoneType) -> Tensor  # noqa
        # type: (Union[Tensor, OptPairTensor], SparseTensor, Size, NoneType) -> Tensor  # noqa
        # type: (Union[Tensor, OptPairTensor], Tensor, Size, bool) -> Tuple[Tensor, Tuple[Tensor, Tensor]]  # noqa
        # type: (Union[Tensor, OptPairTensor], SparseTensor, Size, bool) -> Tuple[Tensor, SparseTensor]  # noqa
        r"""
        Args:
            return_attention_weights (bool, optional): If set to :obj:`True`,
                will additionally return the tuple
                :obj:`(edge_index, attention_weights)`, holding the computed
                attention weights for each edge. (default: :obj:`None`)
        """
        H, C = self.heads, self.out_channels

        x_l: OptTensor = None
        x_r: OptTensor = None
        alpha_l: OptTensor = None
        alpha_r: OptTensor = None
        if isinstance(x, Tensor):
            assert x.dim() == 2, 'Static graphs not supported in `GATConv`.'
            x_l = x_r = self.lin_l(x).view(-1, H, C)
            alpha_l = (x_l * self.att_l).sum(dim=-1)
            alpha_r = (x_r * self.att_r).sum(dim=-1)
        else:
            x_l, x_r = x[0], x[1]
            assert x[0].dim() == 2, 'Static graphs not supported in `GATConv`.'
            x_l = self.lin_l(x_l).view(-1, H, C)
            alpha_l = (x_l * self.att_l).sum(dim=-1)
            if x_r is not None:
                x_r = self.lin_r(x_r).view(-1, H, C)
                alpha_r = (x_r * self.att_r).sum(dim=-1)

        assert x_l is not None
        assert alpha_l is not None

        if self.add_self_loops:
            if isinstance(edge_index, Tensor):
                num_nodes = x_l.size(0)
                if x_r is not None:
                    num_nodes = min(num_nodes, x_r.size(0))
                if size is not None:
                    num_nodes = min(size[0], size[1])
                edge_index, _ = remove_self_loops(edge_index)
                edge_index, _ = add_self_loops(edge_index, num_nodes=num_nodes)
            elif isinstance(edge_index, SparseTensor):
                edge_index = set_diag(edge_index)

        # propagate_type: (x: OptPairTensor, alpha: OptPairTensor)
        out = self.propagate(edge_index, x=(x_l, x_r),
                             alpha=(alpha_l, alpha_r), size=size)

        alpha = self._alpha
        self._alpha = None

        if self.concat:
            out = out.view(-1, self.heads * self.out_channels)
        else:
            out = out.mean(dim=1)

        if self.bias is not None:
            out += self.bias

        if isinstance(return_attention_weights, bool):
            assert alpha is not None
            if isinstance(edge_index, Tensor):
                return out, (edge_index, alpha)
            elif isinstance(edge_index, SparseTensor):
                return out, edge_index.set_value(alpha, layout='coo')
        else:
            return out

    def message(self, x_j: Tensor, alpha_j: Tensor, alpha_i: OptTensor,
                index: Tensor, ptr: OptTensor,
                size_i: Optional[int]) -> Tensor:
        alpha = alpha_j if alpha_i is None else alpha_j + alpha_i
        alpha = F.leaky_relu(alpha, self.negative_slope)
        alpha = softmax(alpha, index, ptr, size_i)
        self._alpha = alpha
        alpha = F.dropout(alpha, p=self.dropout, training=self.training)
        return x_j * alpha.unsqueeze(-1)

    def __repr__(self):
        return '{}({}, {}, heads={})'.format(self.__class__.__name__,
                                             self.in_channels,
                                             self.out_channels, self.heads)


# Data stage randomness.
# There are CUDA randomness that is hard to control.
# They should be adjusted by environment variables.
# Current PyTorch deterministic setup is in beta.
torch.set_deterministic(True)

num_inputs = 7
num_outputs = 5

rng = torch.Generator("cpu")
rng.manual_seed(47)

node_input = torch.zeros(4, num_inputs)
node_input.uniform_(-1, 1, generator=rng)
adj_input = torch.LongTensor([
    [0, 1], [1, 2], [2, 3], [3, 0], [0, 2], [1, 3],
    [1, 0], [2, 1], [3, 2], [0, 3], [2, 0], [3, 1],
]).t()
edge_input = torch.zeros(adj_input.size(1), num_inputs)
edge_input.uniform_(-1, 1, generator=rng)
node_target = torch.zeros(4, num_outputs)
node_target.uniform_(-1, 1, generator=rng)

num_attn_heads = 1
rep = GATConv(
    in_channels=num_inputs, out_channels=num_outputs,
    heads=num_attn_heads,
    concat=True, negative_slope=0.2, dropout=0, add_self_loops=True,
    bias=True,
)
tar = torch_geometric.nn.GATConv(
    in_channels=num_inputs, out_channels=num_outputs,
    heads=num_attn_heads,
    concat=True, negative_slope=0.2, dropout=0, add_self_loops=True,
    bias=True,
)

rep.lin_l.weight.data.fill_(0.5)
rep.lin_r.weight.data.fill_(0.5)
rep.att_l.data.fill_(0.5)
rep.att_r.data.fill_(0.5)
rep.bias.data.fill_(0.5)
tar.lin_l.weight.data.fill_(0.5)
tar.lin_r.weight.data.fill_(0.5)
tar.att_l.data.fill_(0.5)
tar.att_r.data.fill_(0.5)
tar.bias.data.fill_(0.5)

rep_node_output = rep.forward(
    node_input.clone(), adj_input.clone(),
)
tar_node_output = tar.forward(
    node_input.clone(), adj_input.clone(),
)
print(rep_node_output)
print(tar_node_output)
print()

rep_node_output.sum().backward()
tar_node_output.sum().backward()

for name, (rep_grad, tar_grad) in [
    ("weight_l", (
        rep.lin_l.weight.grad, tar.lin_l.weight.grad,
    )),
    ("weight_r", (
        rep.lin_r.weight.grad, tar.lin_r.weight.grad,
    )),
    ("a_l", (
        rep.att_l.grad, tar.att_l.grad,
    )),
    ("a_r", (
        rep.att_r.grad, tar.att_r.grad,
    )),
    ("bias", (
        rep.bias.grad, tar.bias.grad,
    )),
]:
    error = (rep_grad - tar_grad).abs().max().item()
    print(name, error)
```

**Comment by rusty1s** [comment]
Nice findings :) I can't explain why, but those inconsistencies are injected by the `@torch.jit.script` annotation of `torch_geometric.utils.softmax`. After removing this annotation, gradients are exactly equal. I removed it in master for now.

**Comment by alexmirrington** [comment]
To expand on this, I have found a related (but maybe worthy of its own issue) bug where training a model that uses `GATConv` with PyG v1.6.1, saving the weights and then loading them with PyG v1.6.3 yields vastly different outputs. The model I am working on is reasonably complex, but the outputs are different enough that my model that originally achieved 90% accuracy in v1.6.1 gives outputs that equate to approx. 39% accuracy on v1.6.3. I suspect this is due to the GATConv bugfixes in the 1.6.2 release.

I am very short on time at the moment so am not able to provide some minimal code to reproduce the error, but maybe some others are finding similar issues.

**Comment by rusty1s** [comment]
Yes, this is related to the bugfix. If you want to restore the old „bugged“ behavior, you can do `conv.att_r = conv.att_l`.

**Comment by alexmirrington** [comment]
Cheers Matthias, I ended up just reinstalling PyG 1.6.1 for evaluation, just wanted to mention it on this thread in case others are having similar issues to hopefully save them some time!
